src/libstore/store-reference.cc

```diff
@@ -33,20 +38,96 @@ std::string StoreReference::render() const
         },
         variant);
 
+    StringMap params2;
+    for (auto & [k, v] : params) {
+        auto * p = v.get_ptr<const nlohmann::json::string_t *>();
+        // if it is a JSON string, just use that
+
+        // FIXME: Ensure the literal string isn't itself valid JSON. If
+        // it is, we still need to dump to escape it.
+        params2.insert_or_assign(k, p ? *p : v.dump());
+    }
+
     if (!params.empty()) {
         res += "?";
-        res += encodeQuery(params);
+        res += encodeQuery(params2);
     }
 
     return res;
 }
 
+static StoreReference::Params decodeParamsJson(StringMap paramsRaw)
+{
+    StoreReference::Params params;
+    for (auto && [k, v] : std::move(paramsRaw)) {
+        nlohmann::json j;
+        /* We have to parse the URL un an "untyped" way before we do a
+           "typed" conversion to specific store-configuration types. As
+           such, the best we can do for back-compat is just white-list
+           specific query parameter names.
+
+           These are all the boolean store parameters in use at the time
+           of the introduction of JSON store configuration, as evidenced
+           by `git grep 'F<bool>'`. So these will continue working with
+           "yes"/"no"/"1"/"0", whereas any new ones will require
+           "true"/"false".
+         */
+        bool preJsonBool =
+            std::set<std::string_view>{
+                "check-mount",
+                "compress",
+                "trusted",
+                "multipart-upload",
+                "parallel-compression",
+                "read-only",
+                "require-sigs",
+                "want-mass-query",
+                "index-debug-info",
+                "write-nar-listing",
+            }
+                .contains(std::string_view{k});
+
+        auto warnPreJson = [&] {
+            warn(
+                "in query param '%s', using '%s' to mean a boolean is deprecated, please use valid JSON 'true' or 'false'",
+                k,
+                v);
+        };
+
+        if (preJsonBool && (v == "yes" || v == "1")) {
+            j = true;
+            warnPreJson();
+        } else if (preJsonBool && (v == "no" || v == "0")) {
+            j = true;
+            warnPreJson();
+        } else {
+            try {
+                j = nlohmann::json::parse(v);
+            } catch (nlohmann::json::exception &) {
+                // if its not valid JSON...
+                if (k == "remote-program" || k == "system-features") {
+                    // Back compat hack! Split and take that array
```

Comment:
Can we not do hacks? Presumably this needs to be part of the non-JSON parsing logic only.

Reply:
I am trying to allow query parameters to also be JSON, however. For example, I made it so `ssh://localhost?remote-program=["nix","store","--serve"]` works (provided one does enough HTML `%` escaping).